sched/semaphore/spinlock.c

```diff
@@ -446,4 +446,225 @@ void spin_clrbit(FAR volatile cpu_set_t *set, unsigned int cpu,
+{
+  rwlock_t old = atomic_load(lock);
+
+  if (old >= RW_SP_UNLOCKED &&
+      atomic_compare_exchange_strong(lock, &old, old + 1))
+    {
+      SP_DMB();
+      return true;
+    }
+
+  return false;
+}
+
+/****************************************************************************
+ * Name: read_unlock
+ *
+ * Description:
+ *   Release a bit on a non-reentrant spinlock.
+ *
+ * Input Parameters:
+ *   lock - A reference to the spinlock object to unlock.
+ *
+ * Returned Value:
+ *   None.
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+void read_unlock(FAR volatile rwlock_t *lock)
+{
+  DEBUGASSERT(atomic_load(lock) >= RW_SP_READ_LOCKED);
+
+  SP_DMB();
+  atomic_fetch_add(lock, -1);
+  SP_DSB();
+  SP_SEV();
+}
+
+/****************************************************************************
+ * Name: write_lock
+ *
+ * Description:
+ *   If this task does not already hold the spinlock, then loop until the
+ *   spinlock is successfully locked.
+ *
+ *   This implementation is non-reentrant and set all bit on lock to avoid
+ *   readers and writers.
+ *
+ *  The priority of reader is higher than writter if a reader hold the
+ *  lock, a new reader can get its lock but writer can't get this lock.
+ *
+ * Input Parameters:
+ *   lock - A reference to the spinlock object to lock.
+ *
+ * Returned Value:
+ *   None.  When the function returns, the spinlock was successfully locked
+ *   by this CPU.
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+void write_lock(FAR volatile rwlock_t *lock)
+{
+  rwlock_t zero = RW_SP_UNLOCKED;
+
+  while (!atomic_compare_exchange_strong(lock, &zero, RW_SP_WRITE_LOCKED))
+    {
+      SP_DSB();
+      SP_WFE();
+    }
+
+  SP_DMB();
+}
+
+/****************************************************************************
+ * Name: write_trylock
+ *
+ * Description:
+ *   If this task does not already hold the spinlock, then loop until the
+ *   spinlock is successfully locked.
+ *
+ *   This implementation is non-reentrant and set all bit on lock to avoid
+ *   readers and writers.
+ *
+ *  The priority of reader is higher than writter if a reader hold the
+ *  lock, a new reader can get its lock but writer can't get this lock.
+ *
+ * Input Parameters:
+ *   lock - A reference to the spinlock object to lock.
+ *
+ * Returned Value:
+ *   false   - Failure, the spinlock was already locked
+ *   true    - Success, the spinlock was successfully locked
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+bool write_trylock(FAR volatile rwlock_t *lock)
+{
+  rwlock_t zero = RW_SP_UNLOCKED;
+
+  if (!atomic_compare_exchange_strong(lock, &zero, RW_SP_WRITE_LOCKED))
+    {
+      SP_DSB();
```

Comment:
why need in fail case

Reply:
I do't have any special reason.
The logical also can be modified to success case.

Reply:
so should we remove it?

Reply:
Actually if we don't include "arch/spinlock.h", SP_XXX is useless and it will replace with empty line.
So we can remove it.

Reply:
> so should we remove it?

Im not sure what is your it.
But no matter SP_DSB(), SP_DMB()... or using success case, all answers are yes.
We can remove them.